common/src/api/internal/shared.rs

```diff
@@ -58,16 +61,95 @@ pub struct NetworkInterface {
 
 /// An IP address and port range used for source NAT, i.e., making
 /// outbound network connections from guests or services.
-#[derive(
-    Debug, Clone, Copy, Deserialize, Serialize, JsonSchema, PartialEq, Eq, Hash,
-)]
+// Note that `Deserialize` is manually implemented; if you make any changes to
+// the fields of this structure, you must make them to that implementation too.
+#[derive(Debug, Clone, Copy, Serialize, JsonSchema, PartialEq, Eq, Hash)]
 pub struct SourceNatConfig {
     /// The external address provided to the instance or service.
     pub ip: IpAddr,
     /// The first port used for source NAT, inclusive.
-    pub first_port: u16,
+    first_port: u16,
     /// The last port used for source NAT, also inclusive.
-    pub last_port: u16,
+    last_port: u16,
+}
+
+// We implement `Deserialize` manually to add validity checking on the port
+// range.
+impl<'de> Deserialize<'de> for SourceNatConfig {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        use serde::de::Error;
+
+        // The fields of `SourceNatConfigShadow` should exactly match the fields
+        // of `SourceNatConfig`. We're not really using serde's remote derive,
+        // but by adding the attribute we get compile-time checking that all the
+        // field names and types match. (It doesn't check the _order_, but that
+        // should be fine as long as we're using JSON or similar formats.)
+        #[derive(Deserialize)]
+        #[serde(remote = "SourceNatConfig")]
+        struct SourceNatConfigShadow {
+            ip: IpAddr,
+            first_port: u16,
+            last_port: u16,
+        }
+
+        let shadow = SourceNatConfigShadow::deserialize(deserializer)?;
+        SourceNatConfig::new(shadow.ip, shadow.first_port, shadow.last_port)
+            .map_err(D::Error::custom)
+    }
+}
+
+impl SourceNatConfig {
+    /// Construct a `SourceNatConfig` with the given port range, both inclusive.
+    ///
+    /// # Errors
+    ///
+    /// Fails if `(first_port, last_port)` is not aligned to
+    /// [`NUM_SOURCE_NAT_PORTS`].
```

Comment:
So.... I'm wondering if rather than `first_port` and `last_port` this can just take the multiple of `NUM_SOURCE_NAT_PORTS`. Since `NUM_SOURCE_NAT_PORTS` is 2^14 and the number of ports is 2^16, there are only 4 possible values. So this could actually take an enum, say, `SourceNatPortBlock { Block0, Block1, Block2, Block3 }`. That would make this constructor infallible.

Then you can serialize the data in that fashion as well, and deserialization won't require further validation.

Some places may have a port pair, and you can have a constructor that converts that into a `SourceNatPortBlock`, returning the below error. But places that don't have the raw ports won't have to deal with errors.

Not sure how much it increases work for you, maybe in a followup/after r8 if it disrupts this work too much?

Reply:
I like that idea a lot, but yeah I think it's largely orthogonal to the blueprint work. I'll convert this comment into an issue to tackle later.